File: server/schemas/knowledge.py
from typing import Literal

from pydantic import BaseModel, Field

DerivedEdgeType = Literal[
    "is_part_of",
    "assigns_obligation_to",
    "grants_right_to",
    "defines",
]

ExtractedEdgeType = Literal[
    "uses",
    "references",
    "depends_on",
    "supersedes",
    "modifies",
]

AnalysisEdgeType = Literal["contradicts"]

EdgeType = DerivedEdgeType | ExtractedEdgeType | AnalysisEdgeType


class KgParty(BaseModel):
    id: str
    name: str
    role: str = ""
    address: str = ""
    aliases: list[str] = Field(default_factory=list)
    paragraphIds: list[str] = Field(default_factory=list)


class KgClause(BaseModel):
    id: str
    ref: str | None = None
    heading: str = ""
    level: int | None = None
    paragraphIds: list[str] = Field(default_factory=list)


class KgDefinedTerm(BaseModel):
    id: str
    term: str
    definition: str = ""
    definedInClauseId: str | None = None
    paragraphIds: list[str] = Field(default_factory=list)


class _KgDeontic(BaseModel):
    id: str
    action: str = ""
    summary: str
    text: str = ""
    evidenceVerified: bool | None = None
    evidenceSpans: list[str] = Field(default_factory=list)
    burdenPartyId: str | None = None
    benefitPartyId: str | None = None
    clauseId: str | None = None
    deadline: str = ""
    frequency: str = ""
    paragraphIds: list[str] = Field(default_factory=list)


class KgObligation(_KgDeontic):
    pass


class KgRight(_KgDeontic):
    pass


class KgProhibition(_KgDeontic):
    pass


class KgCondition(BaseModel):
    id: str
    trigger: str
    operator: str = ""  # IF | UNLESS | UNTIL | UPON
    gatesId: str | None = None  # statement or clause the condition gates
    paragraphIds: list[str] = Field(default_factory=list)


class KgReference(BaseModel):
    id: str
    name: str
    citation: str = ""
    citedById: str | None = None
    paragraphIds: list[str] = Field(default_factory=list)


class KgValue(BaseModel):
    id: str
    valueType: str = ""  # Currency | Percentage | Duration | Quantity
    amount: str = ""
    unit: str = ""
    quantifiesId: str | None = None
    paragraphIds: list[str] = Field(default_factory=list)


class KgEdge(BaseModel):
    source: str
    target: str
    type: EdgeType
    evidence: str = ""
    paragraphIds: list[str] = Field(default_factory=list)


class KnowledgeGraph(BaseModel):
    parties: list[KgParty] = Field(default_factory=list)
    clauses: list[KgClause] = Field(default_factory=list)
    definedTerms: list[KgDefinedTerm] = Field(default_factory=list)
    obligations: list[KgObligation] = Field(default_factory=list)
    rights: list[KgRight] = Field(default_factory=list)
    prohibitions: list[KgProhibition] = Field(default_factory=list)
    conditions: list[KgCondition] = Field(default_factory=list)
    references: list[KgReference] = Field(default_factory=list)
    values: list[KgValue] = Field(default_factory=list)
    edges: list[KgEdge] = Field(default_factory=list)

File: web/src/types/knowledge.ts
export type DeonticKind = 'obligation' | 'right' | 'prohibition';

export type DerivedEdgeType =
	| 'is_part_of'
	| 'assigns_obligation_to'
	| 'grants_right_to'
	| 'defines';

export type ExtractedEdgeType =
	| 'uses'
	| 'references'
	| 'depends_on'
	| 'supersedes'
	| 'modifies';

export type AnalysisEdgeType = 'contradicts';

export type KgEdgeType = DerivedEdgeType | ExtractedEdgeType | AnalysisEdgeType;

export interface KgParty {
	id: string;
	name: string;
	role: string;
	address: string;
	aliases: string[];
	paragraphIds: string[];
}

export interface KgClause {
	id: string;
	ref: string | null;
	heading: string;
	level: number | null;
	paragraphIds: string[];
}

export interface KgDefinedTerm {
	id: string;
	term: string;
	definition: string;
	definedInClauseId: string | null;
	paragraphIds: string[];
}

export interface KgDeontic {
	id: string;
	action: string;
	summary: string;
	text: string;
	burdenPartyId: string | null;
	benefitPartyId: string | null;
	clauseId: string | null;
	deadline: string;
	frequency: string;
	paragraphIds: string[];
}

export type KgObligation = KgDeontic;
export type KgRight = KgDeontic;
export type KgProhibition = KgDeontic;

export interface KgCondition {
	id: string;
	trigger: string;
	operator: string;
	gatesId: string | null;
	paragraphIds: string[];
}

export interface KgReference {
	id: string;
	name: string;
	citation: string;
	citedById: string | null;
	paragraphIds: string[];
}

export interface KgValue {
	id: string;
	valueType: string;
	amount: string;
	unit: string;
	quantifiesId: string | null;
	paragraphIds: string[];
}

export interface KgEdge {
	source: string;
	target: string;
	type: KgEdgeType;
	evidence: string;
	paragraphIds: string[];
}

export interface KnowledgeGraph {
	parties: KgParty[];
	clauses: KgClause[];
	definedTerms: KgDefinedTerm[];
	obligations: KgObligation[];
	rights: KgRight[];
	prohibitions: KgProhibition[];
	conditions: KgCondition[];
	references: KgReference[];
	values: KgValue[];
	edges: KgEdge[];
}

export interface KnowledgeGraphResponse {
	status: 'success' | 'error';
	documentId: string;
	knowledgeGraph: KnowledgeGraph;
}

export type KgNodeKind =
	| 'party'
	| 'clause'
	| 'definedTerm'
	| 'obligation'
	| 'right'
	| 'prohibition'
	| 'condition'
	| 'reference'
	| 'value';

export interface KgDeonticNode extends KgDeontic {
	kind: DeonticKind;
}

export function deonticNodes(kg: KnowledgeGraph): KgDeonticNode[] {
	return [
		...kg.obligations.map((n) => ({ ...n, kind: 'obligation' as const })),
		...kg.rights.map((n) => ({ ...n, kind: 'right' as const })),
		...kg.prohibitions.map((n) => ({ ...n, kind: 'prohibition' as const })),
	];
}
